geps/gep-1651.md

```diff
@@ -0,0 +1,337 @@
+
+Allow users to configure a Gateway so that it is only routable within
+a specific scope (ie. public/private/cluster)
+
+## Goals
+
+- Define a mechanic to set the routability on a Gateway
+- Provide a default set of routability options
+- Provide a way for vendors to support custom options
+
+## Non-Goals
+
+- Per-request/route scope
+- Not a lightweight service mesh
+
+## Introduction
+
+One of the early feature requests for Knative was the ability to deploy an
+application using Knative's HTTP routing support, but make it only available
+within the cluster. I want to be able to specify both the "cluster"
+(service.namespace.svc) and "external" (service.namespace.example.com).
+Gateways using the same GatewayClass on the cluster, but ensure that the
+"cluster" service is only routable within the cluster. This would greatly
+simplify deployment for users over the instructions we have today.
+
+Likewise another use case is to provide load balancing capabilities within a virtual
+private network. Different IaaS providers offer private load balancers to support
+these use cases.
+
+## API
+
+We propose adding a new `routability` field under the `spec.infrastructure` stanza of a Gateway.
+
+### Predefined Routability Values
+
+Implementations MAY implement the following values for 'routability' and MUST abide by
+their defined semantics.
+
+Value | Scope
+-|-
+`Public`|The address is routable on the public internet
+`Private`|The address is routable inside a private network larger than a single cluster (ie. VPC) and MAY include RFC1918 address space
+`Cluster`|The address is routable inside the [cluster's network](https://kubernetes.io/docs/concepts/cluster-administration/networking/#how-to-implement-the-kubernetes-network-model)
+
+Values can be compared semantically - `Public` has a larger scope than `Private`, while `Private` has a larger scope than `Cluster`.
+
+### Vendor prefixed values
+
+Implementations can define custom 'routability' values by specifying a vendor prefix followed
+by a slash `/` and a custom name ie. `com.example.com/my-routability`.
+
+Comparing vendor prefixed scopes with the pre-defined ones in implementation specific.
+
+### Default Routability
+
+The default value of `routability` is implementation specific. It is RECOMMENDED that
+the default `routability` remains consistent for Gateways with the same
+`gatewayClassName`.
+
+Implementations MAY leave the `spec.infrastructure.routability` property unset and instead
+signal the default routability using the Gateway's `status.addresses`. See 'Status Addresses`
+for more details.
+
+### Mutability
+Implementations MAY prevent end-users from updating the `routability` value of a Gateway. If
+updates are allowed the semantics and behaviour will depend on the underlying implementation.
+
+If a Gateway is mutated but does not support the desired routability it MUST set the conditions
+`Accepted`, `Ready`, `Programmed` to `False` with `Reason` set to `UnsupportedRoutability`.
```

Comment:
```suggestion
`Accepted`, `Ready`, and `Programmed` to `False` with `Reason` set to `UnsupportedRoutability`.
```

@youngnick Any thoughts on what this means for status? I'm unsure if we've ever had a precedent for setting all of these to false in one go + using the same reason for all.

Reply:
Scenario I was thinking about is that if you have a Gateway that's provisioned and `Ready=True` , `Accepted=True`, `Programmed=True` then mutating the routability to something unsupported shouldn't leave conditions in a mixed state.

I'd prefer webhooks to block this mutation but this language is for implementations that don't have that.

Reply:
I wonder if it would make sense to set `Accepted` to false with that reason and the latest generation while leaving both `Porgrammed` and `Ready` true and on the previous generation?

Reply:
> I wonder if it would make sense to set Accepted to false with that reason and the latest generation while leaving both Porgrammed and Ready true and on the previous generation?

How is a consumer supposed to reason about a condition set like that?

Because the generation on the condition doesn't match the `spec.metadata.generation` the values are meaningless and can't safely be used.

Reply:
> So if this happens do we really want implementations to deprovision a Gateway?

The API is declarative - `Programmed=False` in my mind just states that the Gateway spec is not able to be programmed. I don't think that implies previous gateway resources are deprovisioned.

Reply:
That feels like a thing we can put as a _MAY_. "Implementations _may_ choose to leave the old Gateway configuration in place" or something.

Reply:
Feels like that's covered by 
> If updates are allowed the semantics and behaviour will depend on the underlying implementation.

Reply:
> Feels like that's covered by
> 
> > If updates are allowed the semantics and behaviour will depend on the underlying implementation.

I think that only covers when updates to Routability are actually supported. The case I'm concerned about is how we clearly communicate that someone requested an unsupported Routability so we're going to continue to serve the last known good state instead of tearing it down. I like Nick's idea above of saying that "Implementations may choose to leave the old Gateway configuration in place" if the latest generation is invalid.

Finally as one last note here, I think we should avoid anywhere in the spec stating that implementations MUST set the "Ready" condition. I'd prefer removing it entirely from this GEP.

Reply:
@dprotaso I think we need the following changes here:

1. Remove "Ready" from the list of conditions to set
2. Leave some room for implementations to leave a Gateway running with the previous generation of config 

I don't think we need to solve what 2 looks like, we just need to ensure the language in this GEP does not entirely prevent that from happening.